### PR TITLE
#159231621 An endpoint to add person to a center

### PR DESCRIPTION
**Pull request by appcypher**
#### What does this PR do?
- Have an endpoint to add person to center

#### Description of Task to be completed?
- An authenticated user should be able to a person to a center and should get a success message if all went well

#### How should this be manually tested?
- Create a local copy with the following commands: `git fetch origin ft-add-person-to-center-159231621`, then `ft-add-person-to-center-159231621`
- Start the application with the command: `flask run`
- You need to add seed data to database test the api: `flask seed_database`
- If that doesn't work, then you need to recreate the database:
    * Drop database: `dropdb activo`
    * Create database: `createdb activo`
    * Add the tables to the database: `flask db upgrade`
    * Populate database with seed data: `flask seed_database`
- Using postman, send a POST request to: `localhost:5000/api/centers/{<center_id>}/people`
- Run the tests with `pytest`
- **Request body Format**
```javascript
POST /centers/{<center_id>}/people
{
  "name": "Steve Akinyemi"
  "role": "Operations Assistant",
  "email": "email@example.com",
  "imageUrl": "https://res.cloudinary.com/king-more-recipes/image/upload/v1521543073/vijtxydae3bb60encthz.jpg"
}
```
- **Response body format**
```javascript
{ 
  "message": "Steve Akinyemi has been added to LOS center",
  "status": "success"
  "data": {
      "id": "-LIf_HS-Wm6KfPJjg87C"
      "name": "Steve Akinyemi",
      "email": "email@example.com", 
      "imageUrl": "https://res.cloudinary.com/king-more-recipes/image/upload/v1521543073/vijtxydae3bb60encthz.jpg", 
      "centerId": "-LIf_HS-t7tk6WcaTrc56"
      "roleId": "-LIf_HS-1tk6WcQEgcJC"
  }
}
```

#### Any background context you want to provide?
- N/A

#### What are the relevant pivotal tracker stories?
[#159231621](https://www.pivotaltracker.com/story/show/159231621)

#### Postman collection:
[Run in Postman](https://www.getpostman.com/collections/6b7f23715869d36d5bd2)

### CODE REVIEW

File: api/views/user.py [hunk 1]
```diff
@@ -33,3 +36,49 @@ def delete(self, center_id, person_id):
...
+
+    @token_required
+    @validate_id
+    @validate_json_request
+    def post(self, center_id):
+        """Add a user to center"""
+        request_data = request.get_json()
+
+
+        print("Boom")
```
Comment (purpose50) [inline api/views/user.py]: Hi, @appcypher , I understand you added this spaces for readability, but I think it looks good without the spaces also, and owing to the fact that this is not a convention we set I think we should just remove spaces

**Comment by purpose50** [comment]
The PR looks good @appcypher  , but I can see you don't have test for the feature yet, is there any plan towards including the tests ?

File: api/views/user.py [hunk 2]
```diff
@@ -33,3 +36,49 @@ def delete(self, center_id, person_id):
...
+            only=['id', 'name', 'email', 'role_id', 'image_url']
+        ).dump(person).data
+
+        print("B__m")
+
+        return {
+            'status': 'success',
+            'message': SUCCESS_MESSAGES['person_added'].format(person.name, center.name),
+            'data': person_data
+        }
```
Comment (Victoradukwu) [inline api/views/user.py]: @appcypher Good job. A few notes, though.

- fix the changes requested by @purpose50 above.
- resolve the merge conflicts.
- Add tests for this endpoint.
- If this is your final version, remove the debugging print statements in your code.
- It is good practice to notify reviewers when requested changes are treated.

File: api/views/user.py (hunk 1 above)
Comment (frankip) [reply api/views/user.py]: Also remove the print statement

File: api/views/user.py [hunk 3]
```diff
@@ -33,3 +36,49 @@ def delete(self, center_id, person_id):
...
+        request_data = request.get_json()
+
+
+        print("Boom")
+
+        center = Center.get_or_404(center_id)
+
+        print("Baam")
+
+        person = User(
```
Comment (frankip) [inline api/views/user.py]: use the UserSchema class in the validators to create a person, It is clean and the inputs can be validated and also to follow the conventions used through out the project

File: api/views/user.py (hunk 1 above)
Comment (appcypher) [reply api/views/user.py]: Fixed :tada:

File: api/views/user.py (hunk 2 above)
Comment (appcypher) [reply api/views/user.py]: Fixed 🎉

File: api/views/user.py (hunk 3 above)
Comment (appcypher) [reply api/views/user.py]: Fixed 🎉

File: api/utilities/validators/url_validator.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+""" Module with url validator. """
+import re
+
+from marshmallow import ValidationError
+from ..messages.error_messages import serialization_errors
+
+URL_REGEX = re.compile(r"[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)")
```
Comment (GbengaOshinaga) [inline api/utilities/validators/url_validator.py]: @appcypher Could you take a look at your url validation, it doesn't seem to accept even valid urls, like one gotten from cloudinary

Comment (appcypher) [reply api/utilities/validators/url_validator.py]: Hmmm. Will do. 
Can you send cloudinary link you tested it with?

File: api/views/user.py [hunk 5]
```diff
@@ -33,3 +37,73 @@ def delete(self, center_id, person_id):
...
+        # Remove role field from the user data as it cannot be saved into the db
+        del user_data['role']
+
+        # Add the new user data to the database
+        user = User(**user_data)
+        user.save()
+
+        # Schema for serializing back a response
+        user_serialize_schema = UserSchema(
+            only=["id", "name", "email", "role_id", "center_id", "image_url"]
```
Comment (GbengaOshinaga) [inline api/views/user.py]: @appcypher This `only` is not neccessary, since you're selecting everything the schema is going to return regardless

File: api/utilities/validators/url_validator.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+def url_validator(data):
+    """
+    Checks if given string is at least 1 character and only contains characters
+    that make a valid email.
```
Comment (GbengaOshinaga) [inline api/utilities/validators/url_validator.py]: @appcypher I reckon it's `valid url` you wanted to type

File: api/views/user.py [hunk 7]
```diff
@@ -33,3 +37,71 @@ def delete(self, center_id, person_id):
...
+        )
+
+        # Deserialize, validate response data
+        user_deserialize_schema.load_object_into_schema(request_data)
+
+        # Check if center exists.
+        center = Center.get_or_404(center_id)
+
+        # Check if role exists.
+        role_title = request_data["role"]
```
Comment (GbengaOshinaga) [inline api/views/user.py]: @appcypher If the `An Endpoint To Edit A Person` PR is merged before this, you should make use of the `validate_role` function in the `role_validator.py` module

File: api/utilities/validators/url_validator.py (hunk 4 above)
Comment (appcypher) [reply api/utilities/validators/url_validator.py]: Oh. I didn't handle `http/https` prefix. That is fixed now. 🎉 
I will test it exahaustively to see if there are any other edge case.

File: api/utilities/validators/url_validator.py (hunk 6 above)
Comment (appcypher) [reply api/utilities/validators/url_validator.py]: Yep. Fixed

File: api/views/user.py [hunk 8]
```diff
@@ -33,3 +37,73 @@ def delete(self, center_id, person_id):
...
+        role = Role.query.filter_by(deleted=False, title=role_title.title()).first()
+        if role is None:
+            raise ValidationError({
+                'message': serialization_errors["not_found"].format("Role")
+            }, 404)
+
+        # Check if email does not already exist
+        result = User.query.filter_by(deleted=False, email=request_data['email']).first()
```
[the remark this answers is not in the document]
Comment (appcypher) [reply api/views/user.py]: I refactored validate_duplicate to be a little more generic. So that is fixed now. 🎉

File: api/views/user.py (hunk 7 above)
Comment (appcypher) [reply api/views/user.py]: No problem. Will do.

**Comment by appcypher** [comment]
@GbengaOshinaga 
They've all been addressed. I don't know why they still show as `Changes Requested`.
Should I dismiss them?

File: api/utilities/validators/duplicate_validator.py [hunk 9]
```diff
@@ -14,11 +14,13 @@ def validate_duplicate(model, **kwargs):
          kwargs(dict): keyword arguments containing fields to filter query by
     """
 
+    message = message if message else model.__name__
```
Comment (GbengaOshinaga) [inline api/utilities/validators/duplicate_validator.py]: @appcypher This is not necessary, simply make `model.__name__` the default value in the function parameter

File: api/utilities/validators/duplicate_validator.py [hunk 10]
```diff
@@ -5,7 +5,7 @@
 from api.utilities.messages.error_messages import serialization_errors
 
 
-def validate_duplicate(model, **kwargs):
+def validate_duplicate(model, message="", **kwargs):
```
Comment (GbengaOshinaga) [inline api/utilities/validators/duplicate_validator.py]: @appcypher On second thought. considering the fact that the user would not be inputting an email, I reckon the message `User already exists` is more descriptive, and `Email already exists` may be confusing to the user since he/she did not input any email address.

File: tests/test_user_endpoints.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+        assert response.status_code == 400
+        assert response_json['status'] == 'error'
+        assert response_json['errors']['email'][0] == serialization_errors['field_required']
+        assert response_json['errors']['image_url'][0] == serialization_errors['field_required']
+
+
+    def test_add_user_with_no_token(self, client, init_db, new_center):
+        """
+        Should return an 401 status code if a token is provided
```
Comment (Victoradukwu) [inline tests/test_user_endpoints.py]: @appcypher Do you mean `if a token is NOT provided`"

File: tests/test_user_endpoints.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+        response_json = json.loads(response.data.decode(CHARSET))
+
+        assert response.status_code == 401
+        assert response_json['status'] == 'error'
+        assert response_json['message'] == jwt_errors['NO_TOKEN_MSG']
+
+
+    def test_add_user_with_invalid_token(self, client, init_db, new_center):
+        """
+        Should return an 401 status code if the provided token is valid
```
Comment (Victoradukwu) [inline tests/test_user_endpoints.py]: @appcypher Check the first line of this docstring, too, for a typo.

File: tests/test_user_endpoints.py [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+        response_json = json.loads(response.data.decode(CHARSET))
+
+        assert response.status_code == 401
+        assert response_json['status'] == 'error'
+        assert response_json['message'] == jwt_errors['INVALID_TOKEN_MSG']
+
+
+    def test_add_user_with_non_existent_center(self, client, init_db, auth_header):
+        """
+        Should return an 404 status code if the center specified actually exists in the database
```
Comment (Victoradukwu) [inline tests/test_user_endpoints.py]: ...and here too. Just check the docstring for all your test methods.

**Comment by appcypher** [comment]
@Victoradukwu 
Can you send me the urls you tried

File: api/utilities/validators/duplicate_validator.py (hunk 9 above)
Comment (appcypher) [reply api/utilities/validators/duplicate_validator.py]: Fix mention below

File: api/utilities/validators/duplicate_validator.py (hunk 10 above)
Comment (appcypher) [reply api/utilities/validators/duplicate_validator.py]: Changes reverted. Fixed! 🎉

File: tests/test_user_endpoints.py (hunk 11 above)
Comment (appcypher) [reply tests/test_user_endpoints.py]: Fixed 🎉

File: tests/test_user_endpoints.py (hunk 12 above)
Comment (appcypher) [reply tests/test_user_endpoints.py]: Fixed 🎉

File: tests/test_user_endpoints.py (hunk 13 above)
Comment (appcypher) [reply tests/test_user_endpoints.py]: Fixed 🎉

**Comment by Victoradukwu** [comment]
Good job, @appcypher. Now I know the correct image url format. One more thing. The output and input fields should be camelCase, not snake_case. Eg `roleId` instead of `role_id`.

**Comment by appcypher** [comment]
@Victoradukwu input/output fields fixed! 🎉

File: api/utilities/validators/email_validator.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+from ..messages.error_messages import serialization_errors
+
+
+EMAIL_REGEX = re.compile(r"(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])")
+
+
+def email_validator(data):
+    """
+    Checks if given string is at least 1 character and only contains characters
+    that make a valid email. It also checks if email alreaduy exists on specified model.
```
Comment (frankip) [inline api/utilities/validators/email_validator.py]: you have a typo at the _already_

Comment (appcypher) [reply api/utilities/validators/email_validator.py]: Fixed! 🎉

File: api/utilities/validators/email_validator.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+import re
+
+from marshmallow import ValidationError
+from ..messages.error_messages import serialization_errors
+
+
+EMAIL_REGEX = re.compile(r"(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])")
+
+
+def email_validator(data):
```
Comment (SEUNAGBEYE) [inline api/utilities/validators/email_validator.py]: @appcypher Does this validates if an email ends with `@andela.com`?

Comment (appcypher) [reply api/utilities/validators/email_validator.py]: Yeah, it works with `@andela.com` emails.
You can try `steve.akinyemi@andela.com` and it works.

Comment (SEUNAGBEYE) [reply api/utilities/validators/email_validator.py]: @appcypher I meant does it validates if an email is an Andela type email?

Comment (appcypher) [reply api/utilities/validators/email_validator.py]: It validates all types of email. 
I didn't want to make the assumption that only `andela` emails are acceptabe

Comment (SEUNAGBEYE) [reply api/utilities/validators/email_validator.py]: I will merge this because it is not part of your task scope. However, I will create a new story that will fix this.

Comment (appcypher) [reply api/utilities/validators/email_validator.py]: 👍
